core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/ActiveTransactionRegistry.java

```diff
@@ -94,30 +107,38 @@ private ActiveTransactionRegistry() {
 			}
 		}
 
-		activeConnections = initializeCache(timeout, TimeUnit.SECONDS);
-	}
+		primaryCache = CacheBuilder.newBuilder().removalListener(new RemovalListener<UUID, CacheEntry>() {
 
-	private final Cache<UUID, CacheEntry> initializeCache(int timeout, TimeUnit unit) {
-		return CacheBuilder.newBuilder().removalListener(new RemovalListener<UUID, CacheEntry>() {
+			@Override
+			public void onRemoval(RemovalNotification<UUID, CacheEntry> notification) {
+				CacheEntry entry = notification.getValue();
+				try {
+					entry.getConnection().close();
+				}
+				catch (RepositoryException e) {
+					// fall through
+				}
+			}
+		}).build();
+
+		secondaryCache = CacheBuilder.newBuilder().removalListener(new RemovalListener<UUID, CacheEntry>() {
 
 			@Override
 			public void onRemoval(RemovalNotification<UUID, CacheEntry> notification) {
 				if (RemovalCause.EXPIRED.equals(notification.getCause())) {
-					logger.warn("transaction registry item {} removed after expiry", notification.getKey());
-					CacheEntry entry = notification.getValue();
-					try {
-						entry.getConnection().close();
+					final UUID transactionId = notification.getKey();
+					final CacheEntry entry = notification.getValue();
+					synchronized (primaryCache) {
+						if (!entry.getLock().isLocked()) {
+							// no operation active, we can decommission this entry
+							primaryCache.invalidate(transactionId);
+							logger.warn("deregistered expired transaction {}", transactionId);
+						}
```

Comment:
Note to self: as extra insurance we should re-insert the item in the secondary cache if it is still active.